src/analysis/normed/ring/seminorm_constructions.lean

```diff
@@ -0,0 +1,328 @@
+/- In this section we prove Proposition 1.3.2/2 in
+[S. Bosch, U. Güntzer, R. Remmert, *Non-Archimedean Analysis*][bosch-guntzer-remmert]. -/
+
+section ring
+
+variables {R : Type*} [comm_ring R] (c : R) (f : ring_seminorm R) (hc : 0 ≠ f c)
+  (hpm : is_pow_mul f)
+
+/-- For a ring seminorm `f` on `R` and `c ∈ R`, the sequence given by `(f (x * c^n))/((f c)^n)`. -/
+def seminorm_from_const_seq (x : R) : ℕ → ℝ :=
+λ n, (f (x * c^n))/((f c)^n)
+
+lemma seminorm_from_const_nonneg (x : R) (n : ℕ) : 0 ≤ seminorm_from_const_seq c f x n :=
+div_nonneg (map_nonneg f (x * c ^ n)) (pow_nonneg (map_nonneg f c) n)
+
+lemma seminorm_from_const_is_bdd_below (x : R) :
+  bdd_below (set.range (seminorm_from_const_seq c f x)) :=
+begin
+  use 0,
+  rw mem_lower_bounds,
+  rintros r ⟨n, hn⟩,
+  rw ← hn,
+  exact seminorm_from_const_nonneg c f x n,
+end
+
+variable {f}
+
+lemma seminorm_from_const_apply (x : R) (n : ℕ) :
+  seminorm_from_const_seq c f x n = (f (x * c^n))/((f c)^n) :=
+rfl
+
+lemma seminorm_from_const_seq_zero (hf : f 0 = 0) :
+  seminorm_from_const_seq c f 0 = 0 :=
+begin
+  ext n,
+  rw [seminorm_from_const_apply, zero_mul, hf, zero_div],
+  refl,
+end
+
+variable {c}
+
+include hc hpm
+
+lemma seminorm_from_const_seq_one {n : ℕ} (hn : n ≠ 0) :
+  seminorm_from_const_seq c f 1 n = 1 :=
+by rw [seminorm_from_const_apply, one_mul, hpm _ hn, div_self (pow_ne_zero n (ne.symm hc))]
+
+lemma seminorm_from_const_seq_antitone (x : R) :
+  antitone (seminorm_from_const_seq c f x) :=
+begin
+  apply antitone_nat_of_succ_le,
+  have hc_pos : 0 < f c := lt_of_le_of_ne (map_nonneg f _) hc,
+  intros n,
+  simp only [seminorm_from_const_seq],
+  rw [pow_add, ← mul_assoc],
+  apply le_trans ((div_le_div_right (pow_pos hc_pos _)).mpr (map_mul_le_mul f _ _)),
+  rw [hpm c one_ne_zero, mul_div_assoc, pow_add, div_eq_mul_inv, mul_inv_rev, pow_one,
+    ← mul_assoc (f c), mul_inv_cancel (ne.symm hc), div_eq_mul_inv, one_mul],
+end
+
+omit hc hpm
+
+variables (f c)
+
+/-- The real-valued function sending `x ∈ R` to the limit of `(f (x * c^n))/((f c)^n)`. -/
+def seminorm_from_const_def : R → ℝ := λ x, ⨅ n, seminorm_from_const_seq c f x n
+
+include hc hpm
+
+variables {f c}
+
+lemma seminorm_from_const_def_is_limit (x : R) :
+  filter.tendsto ((seminorm_from_const_seq c f x)) filter.at_top
+    (𝓝 (seminorm_from_const_def  c f x)) :=
+tendsto_at_top_is_glb (seminorm_from_const_seq_antitone hc hpm x)
+  (is_glb_cinfi (seminorm_from_const_is_bdd_below c f x))
+
+lemma seminorm_from_const_zero : seminorm_from_const_def c f 0 = 0 :=
+tendsto_nhds_unique (seminorm_from_const_def_is_limit hc hpm 0)
+  (by simpa [seminorm_from_const_seq_zero c (map_zero _)] using tendsto_const_nhds)
+
+lemma seminorm_from_const_is_norm_one_class : seminorm_from_const_def c f 1 = 1 :=
+begin
+  apply tendsto_nhds_unique_of_eventually_eq (seminorm_from_const_def_is_limit hc hpm 1)
+    tendsto_const_nhds,
+  simp only [filter.eventually_eq, filter.eventually_at_top, ge_iff_le],
```

Comment:
This line is painful to read. I guess we are punished from introducing the definition `eventually_eq` which either requires to duplicate a lot of `eventually` api or unfold the definition. Does
```
apply filter.eventually_at_top.mpr,
```
work here? Or maybe with `refine`?

Reply:
Golfed it, thanks!